### ISSUE TITLE
Dockerfile for ros melodic is broken

### ISSUE THREAD

**Issue by razorheadfx** [cartographer-project/cartographer_ros#1334]
Trying to build the current Dockerfile.melodic (at #92a8b81) like so:```docker build -t cartographer_broken -f Dockerfile.melodic .``` breaks.
```
Sending build context to Docker daemon  3.761MB
Step 1/25 : FROM eu.gcr.io/cartographer-141408/cartographer_ros_melodic_base
 ---> 4494cdd37eac
Step 2/25 : ARG CARTOGRAPHER_VERSION=master
 ---> Using cache
 ---> 12fba05dfdf3
Step 3/25 : ARG github_token
 ---> Using cache
 ---> 91a30d1d3d94
Step 4/25 : RUN apt-get update && apt-get install -y sudo && rm -rf /var/lib/apt/lists/*
 ---> Running in 86ce8532ce8f
Get:1 http://archive.ubuntu.com/ubuntu bionic InRelease [242 kB]
Get:2 http://archive.ubuntu.com/ubuntu bionic-updates InRelease [88.7 kB]
Get:3 http://security.ubuntu.com/ubuntu bionic-security InRelease [88.7 kB]
Get:4 http://archive.ubuntu.com/ubuntu bionic-backports InRelease [74.6 kB]
Get:5 http://packages.ros.org/ros/ubuntu bionic InRelease [4669 B]
Get:6 http://archive.ubuntu.com/ubuntu bionic/universe Sources [11.5 MB]
Err:5 http://packages.ros.org/ros/ubuntu bionic InRelease
  The following signatures couldn't be verified because the public key is not available: NO_PUBKEY F42ED6FBAB17C654
Get:7 http://archive.ubuntu.com/ubuntu bionic/main amd64 Packages [1344 kB]
Get:8 http://archive.ubuntu.com/ubuntu bionic/restricted amd64 Packages [13.5 kB]
Get:9 http://archive.ubuntu.com/ubuntu bionic/universe amd64 Packages [11.3 MB]
Get:10 http://security.ubuntu.com/ubuntu bionic-security/universe Sources [188 kB]
Get:11 http://archive.ubuntu.com/ubuntu bionic/multiverse amd64 Packages [186 kB]
Get:12 http://archive.ubuntu.com/ubuntu bionic-updates/universe Sources [325 kB]
Get:13 http://archive.ubuntu.com/ubuntu bionic-updates/universe amd64 Packages [1226 kB]
Get:14 http://archive.ubuntu.com/ubuntu bionic-updates/multiverse amd64 Packages [7239 B]
Get:15 http://archive.ubuntu.com/ubuntu bionic-updates/restricted amd64 Packages [10.8 kB]
Get:16 http://archive.ubuntu.com/ubuntu bionic-updates/main amd64 Packages [862 kB]
Get:17 http://archive.ubuntu.com/ubuntu bionic-backports/main amd64 Packages [2496 B]
Get:18 http://archive.ubuntu.com/ubuntu bionic-backports/universe amd64 Packages [3927 B]
Get:19 http://security.ubuntu.com/ubuntu bionic-security/restricted amd64 Packages [5436 B]
Get:20 http://security.ubuntu.com/ubuntu bionic-security/universe amd64 Packages [720 kB]
Get:21 http://security.ubuntu.com/ubuntu bionic-security/main amd64 Packages [558 kB]
Get:22 http://security.ubuntu.com/ubuntu bionic-security/multiverse amd64 Packages [4169 B]
Reading package lists...
W: GPG error: http://packages.ros.org/ros/ubuntu bionic InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY F42ED6FBAB17C654
E: The repository 'http://packages.ros.org/ros/ubuntu bionic InRelease' is not signed.
The command '/bin/sh -c apt-get update && apt-get install -y sudo && rm -rf /var/lib/apt/lists/*' returned a non-zero code: 100

```

Most likely due to new GPG key deployed for packages.ros.org [in the wake of a ros build farm compromise](https://discourse.ros.org/t/new-gpg-keys-deployed-for-packages-ros-org/9454).

Also: unlike the other Dockerfiles ```Dockerfile.melodic``` does not use the official ros image but something else (which is probably why it broke in this way)
https://github.com/googlecartographer/cartographer_ros/blob/92a8b81a8489dd71a73c6d838c560712311b1da5/Dockerfile.melodic#L14-L16
 ```Dockerfile.lunar``` for reference
https://github.com/googlecartographer/cartographer_ros/blob/92a8b81a8489dd71a73c6d838c560712311b1da5/Dockerfile.lunar#L14-L16

**Comment by razorheadfx** [comment]
Update: apparently there is another issue with this Dockerfile
Deriving from ```ros:melodic``` and trying to build again breaks with:
```
...
Step 8/25 : RUN CARTOGRAPHER_VERSION=$CARTOGRAPHER_VERSION     cartographer_ros/scripts/update_catkin_workspace.sh
 ---> Running in ae006a891d4f

. /opt/ros/${ROS_DISTRO}/setup.sh
#!/usr/bin/env sh
# generated from catkin/cmake/template/setup.sh.in

# Sets various environment variables and sources additional environment hooks.
# It tries it's best to undo changes from a previously sourced setup file before.
# Supported command line options:
# --extend: skips the undoing of changes from a previously sourced setup file
# --local: only considers this workspace but not the chained ones
# In plain sh shell which doesn't support arguments for sourced scripts you can
# set the environment variable `CATKIN_SETUP_UTIL_ARGS=--extend/--local` instead.

# since this file is sourced either use the provided _CATKIN_SETUP_DIR
# or fall back to the destination set at configure time
: ${_CATKIN_SETUP_DIR:=/opt/ros/melodic}
_SETUP_UTIL="$_CATKIN_SETUP_DIR/_setup_util.py"
unset _CATKIN_SETUP_DIR

if [ ! -f "$_SETUP_UTIL" ]; then
  echo "Missing Python script: $_SETUP_UTIL"
  return 22
fi

# detect if running on Darwin platform
_UNAME=`uname -s`
_IS_DARWIN=0
if [ "$_UNAME" = "Darwin" ]; then
  _IS_DARWIN=1
fi
unset _UNAME

# make sure to export all environment variables
export CMAKE_PREFIX_PATH
if [ $_IS_DARWIN -eq 0 ]; then
  export LD_LIBRARY_PATH
else
  export DYLD_LIBRARY_PATH
fi
unset _IS_DARWIN
export PATH
export PKG_CONFIG_PATH
export PYTHONPATH

# remember type of shell if not already set
if [ -z "$CATKIN_SHELL" ]; then
  CATKIN_SHELL=sh
fi

# invoke Python script to generate necessary exports of environment variables
# use TMPDIR if it exists, otherwise fall back to /tmp
if [ -d "${TMPDIR:-}" ]; then
  _TMPDIR="${TMPDIR}"
else
  _TMPDIR=/tmp
fi
_SETUP_TMP=`mktemp "${_TMPDIR}/setup.sh.XXXXXXXXXX"`
unset _TMPDIR
if [ $? -ne 0 -o ! -f "$_SETUP_TMP" ]; then
  echo "Could not create temporary file: $_SETUP_TMP"
  return 1
fi
CATKIN_SHELL=$CATKIN_SHELL "$_SETUP_UTIL" $@ ${CATKIN_SETUP_UTIL_ARGS:-} >> "$_SETUP_TMP"
_RC=$?
if [ $_RC -ne 0 ]; then
  if [ $_RC -eq 2 ]; then
    echo "Could not write the output of '$_SETUP_UTIL' to temporary file '$_SETUP_TMP': may be the disk if full?"
  else
    echo "Failed to run '\"$_SETUP_UTIL\" $@': return code $_RC"
  fi
  unset _RC
  unset _SETUP_UTIL
  rm -f "$_SETUP_TMP"
  unset _SETUP_TMP
  return 1
fi
unset _RC
unset _SETUP_UTIL
. "$_SETUP_TMP"
# prepend folders of workspaces to environment variables
export CMAKE_PREFIX_PATH="/opt/ros/melodic"
export LD_LIBRARY_PATH="/opt/ros/melodic/lib"
export PATH="/opt/ros/melodic/bin:$PATH"
export PKG_CONFIG_PATH="/opt/ros/melodic/lib/pkgconfig"
export PYTHONPATH="/opt/ros/melodic/lib/python2.7/dist-packages"
# found environment hooks in workspaces
export _CATKIN_ENVIRONMENT_HOOKS_COUNT="8"
export _CATKIN_ENVIRONMENT_HOOKS_0="/opt/ros/melodic/etc/catkin/profile.d/1.ros_distro.sh"
export _CATKIN_ENVIRONMENT_HOOKS_0_WORKSPACE="/opt/ros/melodic"
export _CATKIN_ENVIRONMENT_HOOKS_1="/opt/ros/melodic/etc/catkin/profile.d/1.ros_etc_dir.sh"
export _CATKIN_ENVIRONMENT_HOOKS_1_WORKSPACE="/opt/ros/melodic"
export _CATKIN_ENVIRONMENT_HOOKS_2="/opt/ros/melodic/etc/catkin/profile.d/1.ros_package_path.sh"
export _CATKIN_ENVIRONMENT_HOOKS_2_WORKSPACE="/opt/ros/melodic"
export _CATKIN_ENVIRONMENT_HOOKS_3="/opt/ros/melodic/etc/catkin/profile.d/1.ros_python_version.sh"
export _CATKIN_ENVIRONMENT_HOOKS_3_WORKSPACE="/opt/ros/melodic"
export _CATKIN_ENVIRONMENT_HOOKS_4="/opt/ros/melodic/etc/catkin/profile.d/1.ros_version.sh"
export _CATKIN_ENVIRONMENT_HOOKS_4_WORKSPACE="/opt/ros/melodic"
export _CATKIN_ENVIRONMENT_HOOKS_5="/opt/ros/melodic/etc/catkin/profile.d/10.rosbuild.sh"
export _CATKIN_ENVIRONMENT_HOOKS_5_WORKSPACE="/opt/ros/melodic"
export _CATKIN_ENVIRONMENT_HOOKS_6="/opt/ros/melodic/etc/catkin/profile.d/10.roslaunch.sh"
export _CATKIN_ENVIRONMENT_HOOKS_6_WORKSPACE="/opt/ros/melodic"
export _CATKIN_ENVIRONMENT_HOOKS_7="/opt/ros/melodic/etc/catkin/profile.d/99.roslisp.sh"
export _CATKIN_ENVIRONMENT_HOOKS_7_WORKSPACE="/opt/ros/melodic"
rm -f "$_SETUP_TMP"
unset _SETUP_TMP

# source all environment hooks
_i=0
while [ $_i -lt $_CATKIN_ENVIRONMENT_HOOKS_COUNT ]; do
  eval _envfile=\$_CATKIN_ENVIRONMENT_HOOKS_$_i
  unset _CATKIN_ENVIRONMENT_HOOKS_$_i
  eval _envfile_workspace=\$_CATKIN_ENVIRONMENT_HOOKS_${_i}_WORKSPACE
  unset _CATKIN_ENVIRONMENT_HOOKS_${_i}_WORKSPACE
  # set workspace for environment hook
  CATKIN_ENV_HOOK_WORKSPACE=$_envfile_workspace
  . "$_envfile"
  unset CATKIN_ENV_HOOK_WORKSPACE
  _i=$((_i + 1))
done
# generated from ros_environment/env-hooks/1.ros_distro.sh.in

if [ -n "$ROS_DISTRO" -a "$ROS_DISTRO" != "melodic" ]; then
  echo "ROS_DISTRO was set to '$ROS_DISTRO' before. Please make sure that the environment does not mix paths from different distributions."
fi
export ROS_DISTRO=melodic
# generated from ros_environment/env-hooks/1.ros_etc_dir.sh.em

# env variable in installspace
if [ -z "$CATKIN_ENV_HOOK_WORKSPACE" ]; then
  CATKIN_ENV_HOOK_WORKSPACE="/opt/ros/melodic"
fi
export ROS_ETC_DIR="$CATKIN_ENV_HOOK_WORKSPACE/etc/ros"
# generated from ros_environment/env-hooks/1.ros_package_path.sh.em

# python function to generate ROS package path based on all workspaces
PYTHON_CODE_BUILD_ROS_PACKAGE_PATH=$(cat <<EOF
from __future__ import print_function
import os
env_name = 'CMAKE_PREFIX_PATH'
paths = [path for path in os.environ[env_name].split(os.pathsep)] if env_name in os.environ and os.environ[env_name] != '' else []
workspaces = [path for path in paths if os.path.exists(os.path.join(path, '.catkin'))]
paths = []
for workspace in workspaces:
    filename = os.path.join(workspace, '.catkin')
    data = ''
    with open(filename) as f:
        data = f.read()
    if data == '':
        paths.append(os.path.join(workspace, 'share'))
        if os.path.isdir(os.path.join(workspace, 'stacks')):
            paths.append(os.path.join(workspace, 'stacks'))
    else:
        for source_path in data.split(';'):
            paths.append(source_path)
print(os.pathsep.join(paths))
EOF
)
export ROS_PACKAGE_PATH="`/usr/bin/python -c \"$PYTHON_CODE_BUILD_ROS_PACKAGE_PATH\"`"
# generated from ros_environment/env-hooks/1.ros_version.sh.in

export ROS_PYTHON_VERSION=2
# generated from ros_environment/env-hooks/1.ros_version.sh.in

export ROS_VERSION=1
# generated from rosbuild/env-hooks/10.rosbuild.sh.em

# env variables in installspace
if [ -z "$CATKIN_ENV_HOOK_WORKSPACE" ]; then
  CATKIN_ENV_HOOK_WORKSPACE="/opt/ros/melodic"
fi
export ROS_ROOT="$CATKIN_ENV_HOOK_WORKSPACE/share/ros"
# roslaunch/env-hooks/10.roslaunch.sh

if [ ! "$ROS_MASTER_URI" ]; then
  export ROS_MASTER_URI=http://localhost:11311
fi
# generated from roslisp/env-hooks/99.roslisp.sh.in

# python function to generate ROSLISP package directories containing all devel spaces based on all workspaces
PYTHON_CODE_BUILD_ROSLISP_PACKAGE_DIRECTORIES=$(cat <<EOF
from __future__ import print_function
import os
env_name = 'CMAKE_PREFIX_PATH'
paths = [path for path in os.environ[env_name].split(os.pathsep)] if env_name in os.environ and os.environ[env_name] != '' else []
workspaces = [path for path in paths if os.path.exists(os.path.join(path, '.catkin'))]
paths = []
for workspace in workspaces:
    filename = os.path.join(workspace, '.catkin')
    data = ''
    with open(filename) as f:
        data = f.read()
    if data:
        paths.append(os.path.join(workspace, 'share', 'common-lisp'))
print(os.pathsep.join(paths))
EOF
)
export ROSLISP_PACKAGE_DIRECTORIES="`python -c \"$PYTHON_CODE_BUILD_ROSLISP_PACKAGE_DIRECTORIES\"`"
unset _i

unset _CATKIN_ENVIRONMENT_HOOKS_COUNT

cd catkin_ws/src
cartographer_ros/scripts/update_catkin_workspace.sh: 22: cd: can't cd to catkin_ws/src
```

**Comment by MichaelGrupp** [comment]
Looks like https://github.com/googlecartographer/cartographer_ros/issues/1319

**Comment by MichaelGrupp** [comment]
Docker only uses a different base image with pre-built dependencies because there were issues with the CI pipeline a few months ago. Therefore current `Dockerfile.melodic` can't be built directly `FROM ros:melodic`. You can however change it to build directly from the `ros:melodic` base again, this would be essentially reverting this patch: https://github.com/googlecartographer/cartographer_ros/pull/1021 

You could also build `Dockerfile.melodic.base` yourself and just change the `FROM` statement in `Dockerfile.melodic` to include your freshly built base image instead of the outdated one from Google.

**Comment by bochen87** [comment]
should be fixed by #1484
